src/emulator.js

```diff
@@ -78,59 +88,69 @@ export class Emulator {
+    this.process = spawn("flow", [
+      "emulator",
+      "--verbose",
+      `--log-format=JSON`,
+      `--admin-port=${port}`,
+      `--port=${grpc}`,
+      flags,
+    ]);
     this.logProcessor = (item) => item;
 
     return new Promise((resolve, reject) => {
       let internalId;
-      const checkLiveness = async function() {
+      const checkLiveness = async function () {
         try {
           await send(build([getBlock(false)])).then(decode);
-
           clearInterval(internalId);
           this.initialized = true;
           resolve(true);
-        } catch (err) {}  // eslint-disable-line no-unused-vars, no-empty
-      }
+        } catch (err) {} // eslint-disable-line no-unused-vars, no-empty
+      };
       internalId = setInterval(checkLiveness, 100);
 
-      this.process.stdout.on("data", (data) => {
-        // const buf = this.parseDataBuffer(data);
-
-        if (this.filters.length > 0) {
-          for (let i = 0; i < this.filters.length; i++) {
-            const filter = this.filters[i];
-            if (data.includes(`${filter}`)) {
-              // TODO: use this.log to output string with this.logProcessor and type
-              // TODO: Fix output colors: https://stackoverflow.com/questions/9781218/how-to-change-node-jss-console-font-color
-              // this.log(`LOG: ${data.toString().replace(/\\x1b\[1;34m/, "\x1b[36m")}`);
-              this.log(`LOG: ${data}`);
-              break;
-            }
+      this.process.stdout.on("data", (buffer) => {
+        const data = this.parseDataBuffer(buffer);
+
+        if (Array.isArray(data)) {
+          const filtered = data.filter((item) => {
+            return this.filters.includes(item.level);
+          });
+          for (let i = 0; i < filtered; i++) {
+            const { level = "log", msg } = data[i];
+            this.log(`${level.toUpperCase()}: ${msg}`);
           }
         } else {
-          this.log(`LOG: ${data}`);
-        }
-        if (data.includes("Starting HTTP server")) {
-          this.log("EMULATOR IS UP! Listening for events!");
+          if (this.filters.includes(data.level)) {
+            const { level, msg } = data;
+            this.log(`${level.toUpperCase()}: ${msg}`);
+            if (data.msg.includes("Starting HTTP server")) {
+              this.log("EMULATOR IS UP! Listening for events!");
+            }
+          }
         }
       });
 
-      this.process.stderr.on("data", (data) => {
-        this.log(`ERROR: ${data}`, "error");
+      this.process.stderr.on("data", (buffer) => {
+        const { message } = this.parseDataBuffer(buffer);
+
+        this.log(`EMULATOR ERROR: ${message}`, "error");
         this.initialized = false;
         clearInterval(internalId);
         reject();
       });
 
       this.process.on("close", (code) => {
-        this.log(`emulator exited with code ${code}`);
+        if (this.filters.includes("service")) {
```

Comment:
is this filtering too broad?

Reply:
There is 0 messages emitted with this level by emulator natively.
I was just thinking about the way to filter messages like `process exited with code 2`